xdp/src/tx_loop.rs

```diff
@@ -0,0 +1,278 @@
+#![allow(clippy::arithmetic_side_effects)]
+
+use {
+    crate::{
+        device::{NetworkDevice, QueueId},
+        netlink::MacAddress,
+        packet::{
+            write_eth_header, write_ip_header, write_udp_header, ETH_HEADER_SIZE, IP_HEADER_SIZE,
+            UDP_HEADER_SIZE,
+        },
+        route::Router,
+        set_cpu_affinity,
+        socket::{Socket, Tx, TxRing},
+        umem::{Frame as _, PageAlignedMemory, SliceUmem, SliceUmemFrame, Umem as _},
+    },
+    caps::{CapSet, Capability},
+    crossbeam_channel::{Receiver, Sender, TryRecvError},
+    libc::{sysconf, _SC_PAGESIZE},
+    std::{
+        net::{IpAddr, SocketAddr},
+        thread,
+        time::Duration,
+    },
+};
+
+pub fn tx_loop<T: AsRef<[u8]>>(
+    dev: &NetworkDevice,
+    src_port: u16,
+    queue_id: QueueId,
+    zero_copy: bool,
+    cpu_id: usize,
+    receiver: Receiver<(Vec<SocketAddr>, T)>,
+    drop_sender: Sender<(Vec<SocketAddr>, T)>,
+) {
+    log::info!(
+        "starting xdp loop on {} queue {queue_id:?} cpu {cpu_id}",
+        dev.name()
+    );
+    let src_mac = dev.mac_addr().unwrap();
+    let src_ip = dev.ipv4_addr().unwrap();
+
+    // some drivers require frame_size=page_size
+    let frame_size = unsafe { sysconf(_SC_PAGESIZE) } as usize;
+    const FRAME_COUNT: usize = 4096;
+
+    // try to allocate huge pages first, then fall back to regular pages
+    const HUGE_2MB: usize = 2 * 1024 * 1024;
+    let mut memory =
+        PageAlignedMemory::alloc_with_page_size(frame_size, FRAME_COUNT, HUGE_2MB, true)
+            .or_else(|_| {
+                log::warn!("huge page alloc failed, falling back to regular page size");
+                PageAlignedMemory::alloc(frame_size, FRAME_COUNT)
+            })
+            .unwrap();
+    let umem = SliceUmem::new(&mut memory, frame_size as u32).unwrap();
+
+    // A nice round number. This is the size used by kernel selftests, so likely to work with all
+    // drivers.
+    const RING_SIZE: usize = 2048;
+    let Ok((mut socket, tx)) = Socket::tx(
+        dev.open_queue(queue_id),
+        umem,
+        zero_copy,
+        RING_SIZE,
+        RING_SIZE,
+    ) else {
+        panic!("failed to create AF_XDP socket on queue {queue_id:?}");
+    };
+
+    let umem = socket.umem();
+    let Tx {
+        // this is where we'll queue frames
+        ring,
+        // this is where we'll get completion events once frames have been picked up by the NIC
+        mut completion,
+    } = tx;
+    let mut ring = ring.unwrap();
+
+    // get the routing table from netlink
+    let router = Router::new().expect("failed to create router");
+
+    // each queue is bound to its own CPU core
+    set_cpu_affinity([cpu_id]).unwrap();
+
+    // we don't need CAP_NET_ADMIN anymore
+    caps::drop(None, CapSet::Effective, Capability::CAP_NET_ADMIN).unwrap();
+
+    // How long we sleep waiting to receive shreds from the channel.
+    const RECV_TIMEOUT: Duration = Duration::from_nanos(100);
```

Comment:
I think non-RT kernels can not realistically sleep for such a short interval of time...

Reply:
Oops, this is meant to be 1000 not 100

Reply:
done